File: src/lib/topPicks.js
/**
 * Today's Top 3 Picks — strict A+ filter on top of findPotentialBreakouts.
 *
 * The funnel:
 *   findPotentialBreakouts (~30 stocks)
 *     ↓ require volume confirmation + earnings safety
 *     ↓ score remaining stocks 0-100 across 7 factors
 *     ↓ keep only those scoring ≥50
 *     ↓ take top 3 by score
 *   = up to 3 A+ setups for the day, sometimes zero.
 *
 * Zero picks is a feature, not a bug. Disciplined trading means
 * waiting for high-quality setups instead of forcing trades.
 */

import { findPotentialBreakouts } from "./screener";

const MIN_SCORE = 50;
const MAX_PICKS = 3;
const EARNINGS_BLACKOUT_DAYS = 7;

function daysUntilEarnings(earningsDate) {
  if (!earningsDate) return Infinity;
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  const ed = new Date(earningsDate);
  return Math.round((ed - today) / (1000 * 60 * 60 * 24));
}

function scoreVolumeSurge(volumeSurge) {
  if (volumeSurge === "surge") return { points: 20, label: "SURGE volume" };
  if (volumeSurge === "rising") return { points: 10, label: "Rising volume" };
  return { points: 0, label: "No volume confirmation" };
}

function scoreRS(rsScore) {
  if (rsScore >= 90) return { points: 25, label: `RS ${rsScore} (A+)` };
  if (rsScore >= 85) return { points: 20, label: `RS ${rsScore} (A)` };
  return { points: 10, label: `RS ${rsScore}` };
}

function scoreSector(strengthScore) {
  if (strengthScore == null) return { points: 0, label: "Sector unknown" };
  if (strengthScore >= 80) return { points: 15, label: `Sector ${strengthScore} (leading)` };
  if (strengthScore >= 60) return { points: 10, label: `Sector ${strengthScore} (strong)` };
  if (strengthScore >= 40) return { points: 5, label: `Sector ${strengthScore} (mid)` };
  return { points: 0, label: `Sector ${strengthScore} (weak)` };
}

function scoreEPS(epsGrowth) {
  if (epsGrowth == null) return { points: 0, label: "EPS unknown" };
  if (epsGrowth >= 40) return { points: 15, label: `EPS +${Math.round(epsGrowth)}% (explosive)` };
  if (epsGrowth >= 25) return { points: 8, label: `EPS +${Math.round(epsGrowth)}% (strong)` };
  if (epsGrowth > 0) return { points: 3, label: `EPS +${Math.round(epsGrowth)}%` };
  return { points: 0, label: `EPS ${Math.round(epsGrowth)}% (weak)` };
}

function scoreSales(salesGrowth) {
  if (salesGrowth == null) return { points: 0, label: "Sales unknown" };
  if (salesGrowth >= 40) return { points: 10, label: `Sales +${Math.round(salesGrowth)}%` };
  if (salesGrowth >= 25) return { points: 5, label: `Sales +${Math.round(salesGrowth)}%` };
  if (salesGrowth > 0) return { points: 2, label: `Sales +${Math.round(salesGrowth)}%` };
  return { points: 0, label: `Sales ${Math.round(salesGrowth)}% (weak)` };
}

function scorePivot(distToPivotPct) {
  if (distToPivotPct == null) return { points: 0, label: "Pivot unknown" };
  if (distToPivotPct <= 0.5) return { points: 10, label: `${distToPivotPct.toFixed(1)}% to pivot (imminent)` };
  if (distToPivotPct <= 1) return { points: 7, label: `${distToPivotPct.toFixed(1)}% to pivot (close)` };
  return { points: 4, label: `${distToPivotPct.toFixed(1)}% to pivot` };
}

function scoreVCP(vcpStatus) {
  if (vcpStatus === "tight") return { points: 5, label: "VCP tight" };
  if (vcpStatus === "loose") return { points: 2, label: "VCP loose" };
  return { points: 0, label: null };
}

/**
 * Score a single ticker 0-100 across all factors.
 * Returns { score, breakdown: [...{points, label}], grade }.
 */
export function computeSetupScore(ticker, sectorMap) {
  const sectorStrength = sectorMap[ticker.sector]?.strengthScore;

  const breakdown = {
    rs: scoreRS(ticker.rsScore),
    volume: scoreVolumeSurge(ticker.volumeSurge),
    sector: scoreSector(sectorStrength),
    eps: scoreEPS(ticker.epsGrowthYoY),
    sales: scoreSales(ticker.salesGrowthYoY),
    pivot: scorePivot(ticker.distToPivotPct),
    vcp: scoreVCP(ticker.vcpStatus),
  };

  const score = Object.values(breakdown).reduce((sum, b) => sum + b.points, 0);

  let grade;
  if (score >= 80) grade = "A+";
  else if (score >= 65) grade = "A";
  else if (score >= 50) grade = "B+";
  else if (score >= 35) grade = "B";
  else grade = "C";

  return { score, grade, breakdown };
}

/**
 * Find today's top picks — strictest filter of all.
 *
 * @param {Array} tickers — full ticker list from mockData
 * @param {Array} sectors — sector list (for strength lookup)
 * @returns {{ picks: Array, totalCandidates: number, rejectedReasons: Object }}
 *   picks: array of { ticker, score, grade, breakdown, daysToEarnings }
 *   totalCandidates: # of stocks in initial Potential Breakouts pool
 *   rejectedReasons: counts of why candidates were filtered out
 */
export function findTopPicks(tickers, sectors) {
  const sectorMap = {};
  for (const s of sectors || []) {
    sectorMap[s.name] = s;
  }

  // Step 1: start with the existing Potential Breakouts list
  const candidates = findPotentialBreakouts(tickers);

  const rejectedReasons = {
    noVolumeConfirmation: 0,
    earningsTooSoon: 0,
    scoreTooLow: 0,
  };

  // Step 2: apply hard filters
  const eligible = [];
  for (const t of candidates) {
    if (!t.volumeSurge || t.volumeSurge === "none") {
      rejectedReasons.noVolumeConfirmation++;
      continue;
    }
    const dte = daysUntilEarnings(t.earningsDate);
    if (dte < EARNINGS_BLACKOUT_DAYS) {
      rejectedReasons.earningsTooSoon++;
      continue;
    }
    eligible.push({ ticker: t, daysToEarnings: dte });
  }

  // Step 3: score and rank
  const scored = eligible.map(({ ticker, daysToEarnings }) => {
    const { score, grade, breakdown } = computeSetupScore(ticker, sectorMap);
    return { ticker, score, grade, breakdown, daysToEarnings };
  });

  // Step 4: minimum score gate
  const passed = scored.filter(s => {
    if (s.score < MIN_SCORE) {
      rejectedReasons.scoreTooLow++;
      return false;
    }
    return true;
  });

  // Step 5: top N by score
  passed.sort((a, b) => b.score - a.score);
  const picks = passed.slice(0, MAX_PICKS);

  return {
    picks,
    totalCandidates: candidates.length,
    rejectedReasons,
  };
}

export { MIN_SCORE, MAX_PICKS, EARNINGS_BLACKOUT_DAYS };

File: scripts/top_picks.py
"""
Python port of src/lib/topPicks.js — strict A+ filter on top of Potential Breakouts.

Used by fetch_data.py to record which tickers qualified as Top Picks each day,
so the Performance Tracker can compare Top Picks vs. broader Breakouts over time.

Hard filters (must pass):
  - Volume confirmation (surge / rising)
  - No earnings within 7 days

Score (0-100):
  - RS:        25 (≥90), 20 (≥85), 10 (≥80)
  - Volume:    20 (surge), 10 (rising)
  - Sector:    15 (≥80), 10 (≥60), 5 (≥40)
  - EPS YoY:   15 (≥40%), 8 (≥25%), 3 (>0)
  - Sales YoY: 10 (≥40%), 5 (≥25%), 2 (>0)
  - Pivot:     10 (≤0.5%), 7 (≤1%), 4 (≤2%)
  - VCP:       5 (tight), 2 (loose)

Min score 50 to qualify; up to 3 picks per day.
"""

from datetime import datetime, timezone

MIN_SCORE = 50
MAX_PICKS = 3
EARNINGS_BLACKOUT_DAYS = 7


def days_until_earnings(earnings_date_str):
    if not earnings_date_str:
        return float("inf")
    try:
        ed = datetime.strptime(earnings_date_str, "%Y-%m-%d")
        today = datetime.utcnow()
        return (ed - today).days
    except Exception:
        return float("inf")


def score_volume(volume_surge):
    if volume_surge == "surge":  return 20
    if volume_surge == "rising": return 10
    return 0


def score_rs(rs):
    if rs is None:    return 0
    if rs >= 90:      return 25
    if rs >= 85:      return 20
    return 10


def score_sector(strength):
    if strength is None: return 0
    if strength >= 80:   return 15
    if strength >= 60:   return 10
    if strength >= 40:   return 5
    return 0


def score_eps(eps):
    if eps is None: return 0
    if eps >= 40:   return 15
    if eps >= 25:   return 8
    if eps > 0:     return 3
    return 0


def score_sales(sales):
    if sales is None: return 0
    if sales >= 40:   return 10
    if sales >= 25:   return 5
    if sales > 0:     return 2
    return 0


def score_pivot(dist_pct):
    if dist_pct is None: return 0
    if dist_pct <= 0.5:  return 10
    if dist_pct <= 1:    return 7
    return 4


def score_vcp(vcp):
    if vcp == "tight": return 5
    if vcp == "loose": return 2
    return 0


def grade_from_score(score):
    if score >= 80: return "A+"
    if score >= 65: return "A"
    if score >= 50: return "B+"
    if score >= 35: return "B"
    return "C"


def compute_tags(ticker):
    """Non-blocking observation tags — mirror of computeTags() in
    src/lib/topPicks.js. These do NOT affect the score; they're recorded in
    performance_data.json so we can later measure whether picks carrying a
    given trait outperform or underperform.

      extended    — price stretched >=15% above its 20-day MA
      eps_extreme — EPS YoY > 300% (likely low-base / one-time — verify quality)
    """
    tags = []
    sma20 = ticker.get("sma20")
    price = ticker.get("price")
    if sma20 and sma20 > 0 and price:
        ext = (price - sma20) / sma20 * 100
        if ext >= 15:
            tags.append("extended")
    eps = ticker.get("epsGrowthYoY")
    if eps is not None and eps > 300:
        tags.append("eps_extreme")
    return tags


def is_potential_breakout(ticker):
    """Mirror of src/lib/screener.js findPotentialBreakouts gate.
    Stage 2 is already enforced by the upstream filter, so we only check
    the additional RS≥80 + within 2% of pivot rules here."""
    rs = ticker.get("rsScore", 0)
    if rs < 80:
        return False
    dist = ticker.get("distToPivotPct")
    if dist is None or dist > 2:
        return False
    return True


def compute_top_picks(tickers, sectors):
    """Returns list of dicts: { ticker, score, grade, days_to_earnings }
    sorted by score descending, up to MAX_PICKS entries."""
    sector_strength = {s["name"]: s.get("strengthScore") for s in (sectors or [])}

    candidates = [t for t in tickers if is_potential_breakout(t)]

    eligible = []
    for t in candidates:
        if score_volume(t.get("volumeSurge", "none")) == 0:
            continue
        dte = days_until_earnings(t.get("earningsDate"))
        if dte < EARNINGS_BLACKOUT_DAYS:
            continue
        eligible.append((t, dte))

    scored = []
    for t, dte in eligible:
        score = (
            score_rs(t.get("rsScore"))
            + score_volume(t.get("volumeSurge"))
            + score_sector(sector_strength.get(t.get("sector")))
            + score_eps(t.get("epsGrowthYoY"))
            + score_sales(t.get("salesGrowthYoY"))
            + score_pivot(t.get("distToPivotPct"))
            + score_vcp(t.get("vcpStatus"))
        )
        if score >= MIN_SCORE:
            scored.append({
                "ticker": t["ticker"],
                "price": t.get("price"),
                "score": score,
                "grade": grade_from_score(score),
                "daysToEarnings": dte if dte != float("inf") else None,
                "tags": compute_tags(t),
            })

    scored.sort(key=lambda x: x["score"], reverse=True)
    return scored[:MAX_PICKS]
